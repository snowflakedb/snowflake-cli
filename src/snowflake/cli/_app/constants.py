# Copyright (c) 2024 Snowflake Inc.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
# http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

from __future__ import annotations

from typing import Literal

PARAM_APPLICATION_NAME: Literal["snowcli"] = "snowcli"

# This is also defined on server side. Changing this parameter would require
# a change in https://github.com/snowflakedb/snowflake
INTERNAL_APPLICATION_NAME: Literal["SNOWFLAKE_CLI"] = "SNOWFLAKE_CLI"

# Authenticator types
AUTHENTICATOR_WORKLOAD_IDENTITY: Literal["WORKLOAD_IDENTITY"] = "WORKLOAD_IDENTITY"
AUTHENTICATOR_SNOWFLAKE_JWT: Literal["SNOWFLAKE_JWT"] = "SNOWFLAKE_JWT"
AUTHENTICATOR_USERNAME_PASSWORD_MFA: Literal[
    "username_password_mfa"
] = "username_password_mfa"
AUTHENTICATOR_OAUTH_AUTHORIZATION_CODE: Literal[
    "OAUTH_AUTHORIZATION_CODE"
] = "OAUTH_AUTHORIZATION_CODE"
