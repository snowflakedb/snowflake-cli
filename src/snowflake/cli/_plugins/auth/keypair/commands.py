from pathlib import Path

import typer
from snowflake.cli._plugins.auth.keypair.manager import AuthManager, PublicKeyProperty
from snowflake.cli.api.commands.flags import SecretTypeParser
from snowflake.cli.api.commands.snow_typer import SnowTyperFactory
from snowflake.cli.api.output.types import (
    CollectionResult,
    CommandResult,
    MessageResult,
    SingleQueryResult,
)
from snowflake.cli.api.secret import SecretType
from snowflake.cli.api.secure_path import SecurePath

app = SnowTyperFactory(
    name="keypair",
    help="Manages authentication.",
)


KEY_PAIR_DEFAULT_PATH = "~/.ssh"


def _show_connection_name_prompt(ctx: typer.Context, value: str):
    for param in ctx.command.params:
        if param.name == "connection_name":
            if value:
                param.prompt = "Enter connection name"
            break
    return value


_new_connection_option = typer.Option(
    True,
    help="Create a new connection.",
    prompt="Create a new connection?",
    callback=_show_connection_name_prompt,
    show_default=False,
    hidden=True,
)


_connection_name_option = typer.Option(
    None,
    help="The new connection name.",
    prompt=False,
    show_default=False,
    hidden=True,
)


_key_length_option = typer.Option(
    2048,
    "--key-length",
    help="The RSA key length.",
    prompt="Enter key length",
)


_output_path_option = typer.Option(
    KEY_PAIR_DEFAULT_PATH,
    "--output-path",
    help="The output path for private and public keys",
    prompt="Enter output path",
)


_private_key_passphrase_option = typer.Option(
    "",
    "--private-key-passphrase",
    help="The private key passphrase.",
    click_type=SecretTypeParser(),
    prompt="Enter private key passphrase",
    hide_input=True,
    show_default=False,
)


@app.command("setup", requires_connection=True)
def setup(
    new_connection: bool = _new_connection_option,
    connection_name: str = _connection_name_option,
    key_length: int = _key_length_option,
    output_path: Path = _output_path_option,
    private_key_passphrase: SecretType = _private_key_passphrase_option,
    **options,
):
    """
    Generates the key pair, sets the public key for the user in Snowflake and creates or updates the connection.
    """
    AuthManager().setup(
        connection_name=connection_name,
        key_length=key_length,
        output_path=SecurePath(output_path),
        private_key_passphrase=private_key_passphrase,
    )
    return MessageResult(f"Setup completed.")


@app.command("rotate", requires_connection=True)
def rotate(
    new_connection: bool = _new_connection_option,
    connection_name: str = _connection_name_option,
    key_length: int = _key_length_option,
    output_path: Path = _output_path_option,
    private_key_passphrase: SecretType = _private_key_passphrase_option,
    **options,
):
    """
    Rotates the key for the connection. Generates the key pair, sets the public key for the user in Snowflake and creates or updates the connection.
    """
    AuthManager().rotate(
        connection_name=connection_name,
        key_length=key_length,
        output_path=SecurePath(output_path),
        private_key_passphrase=private_key_passphrase,
    )
    return MessageResult(f"Rotate completed.")


@app.command("list", requires_connection=True)
def list_keys(**options) -> CommandResult:
    """
    Lists the public keys set for the user.
    """
    return CollectionResult(AuthManager().list_keys())


@app.command("remove", requires_connection=True)
def remove(
    public_key_property: PublicKeyProperty = typer.Option(
        ...,
        "--key-id",
        help=f"local path to template directory or URL to git repository with templates.",
        show_default=False,
    ),
    **options,
):
    """
    Removes the public key for the user.
    """
    return SingleQueryResult(
        AuthManager().remove_public_key(public_key_property=public_key_property)
    )


@app.command("status", requires_connection=True)
def status(**options):
    """
    Verifies the key pair configuration and tests the connection.
    """
    AuthManager().status()
    return MessageResult("Status check completed.")
