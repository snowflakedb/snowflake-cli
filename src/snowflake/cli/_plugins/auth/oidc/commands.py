# Copyright (c) 2024 Snowflake Inc.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
# http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

import typer
from snowflake.cli._app.auth.oidc_providers import (
    OidcProviderTypeWithAuto,
)
from snowflake.cli._plugins.auth.oidc.manager import OidcManager
from snowflake.cli.api.commands.snow_typer import SnowTyperFactory
from snowflake.cli.api.output.types import MessageResult

app = SnowTyperFactory(
    name="oidc",
    help="Manages OIDC authentication.",
)


AutoProviderTypeOption = typer.Option(
    OidcProviderTypeWithAuto.AUTO.value,
    "--type",
    help=f"Type of OIDC provider to use",
    show_default=False,
)


@app.command("read-token", requires_connection=False)
def read_token(
    _type: OidcProviderTypeWithAuto = AutoProviderTypeOption,
    **options,
):
    """
    Reads OIDC token based on the specified type.
    Use 'auto' to auto-detect available providers.
    """
    result = OidcManager().read_token(provider_type=_type)
    return MessageResult(result)
