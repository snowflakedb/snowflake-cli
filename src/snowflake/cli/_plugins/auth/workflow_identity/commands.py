# Copyright (c) 2024 Snowflake Inc.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
# http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

import typer
from snowflake.cli._plugins.auth.workflow_identity.manager import (
    WorkflowIdentityManager,
)
from snowflake.cli.api.commands.snow_typer import SnowTyperFactory
from snowflake.cli.api.output.types import (
    CommandResult,
    MessageResult,
    SingleQueryResult,
)

# Main workflow-identity app
app = SnowTyperFactory(
    name="workflow-identity",
    help="Manages GitHub workflow identity federation authentication.",
)


def _setup_command(
    github_repository: str = typer.Option(
        ...,
        "--github-repository",
        help="GitHub repository in format 'owner/repo'",
        prompt="Enter GitHub repository (owner/repo)",
    ),
    **options,
):
    """
    Sets up GitHub workflow identity federation for authentication.
    """
    WorkflowIdentityManager().setup(github_repository=github_repository)
    return MessageResult("GitHub workflow identity federation setup completed.")


def _status_command(**options) -> CommandResult:
    """
    Shows the status of GitHub workflow identity federation configuration.
    """
    result = WorkflowIdentityManager().status()
    return MessageResult(f"GitHub workflow identity federation status: {result}")


def _remove_command(**options) -> CommandResult:
    """
    Removes the GitHub workflow identity federation configuration.
    """
    return SingleQueryResult(WorkflowIdentityManager().remove())


# Register commands on both apps
@app.command("setup", requires_connection=True)
def setup(
    github_repository: str = typer.Option(
        ...,
        "--github-repository",
        help="GitHub repository in format 'owner/repo'",
        prompt="Enter GitHub repository (owner/repo)",
    ),
    **options,
):
    return _setup_command(github_repository=github_repository, **options)


@app.command("status", requires_connection=True)
def status(**options):
    return _status_command(**options)


@app.command("remove", requires_connection=True)
def remove(**options):
    return _remove_command(**options)
