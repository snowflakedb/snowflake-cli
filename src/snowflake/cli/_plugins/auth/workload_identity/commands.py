# Copyright (c) 2024 Snowflake Inc.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
# http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

import typer
from snowflake.cli._app.auth.oidc_providers import (
    OidcProviderType,
)
from snowflake.cli._plugins.auth.workload_identity.manager import (
    WorkloadIdentityManager,
)
from snowflake.cli.api.commands.snow_typer import SnowTyperFactory
from snowflake.cli.api.output.types import MessageResult

app = SnowTyperFactory(
    name="workload-identity",
    help="Manages GitHub workload identity federation authentication.",
)


@app.command("setup", requires_connection=True)
def setup(
    github_repository: str = typer.Option(
        ...,
        "--github-repository",
        show_default=False,
        help="GitHub repository in format 'owner/repo'",
        prompt="Enter GitHub repository (owner/repo)",
    ),
    **options,
):
    """
    Sets up GitHub workload identity federation for authentication.
    """
    result = WorkloadIdentityManager().setup(github_repository=github_repository)
    return MessageResult(result)


@app.command("read", requires_connection=False)
def read(
    _type: str = typer.Option(
        "auto",
        "--type",
        help=f"Type of read operation to perform (e.g., '{OidcProviderType.GITHUB.value}', 'auto')",
    ),
    **options,
):
    """
    Reads OIDC token based on the specified type.
    Use 'auto' to auto-detect available providers.
    """
    result = WorkloadIdentityManager().read(provider_type=_type)
    return MessageResult(result)
