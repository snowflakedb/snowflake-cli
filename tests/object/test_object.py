# Copyright (c) 2024 Snowflake Inc.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
# http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
from tempfile import NamedTemporaryFile
from unittest import mock

import pytest
from click import ClickException
from snowflake.cli._plugins.object.commands import _scope_validate
from snowflake.cli.api.constants import OBJECT_TO_NAMES, SUPPORTED_OBJECTS

from tests.testing_utils.result_assertions import assert_that_result_is_usage_error


@mock.patch("snowflake.connector.connect")
@pytest.mark.parametrize(
    "object_type, expected",
    [
        ("compute-pool", "compute pools"),
        ("network-rule", "network rules"),
        ("database", "databases"),
        ("function", "functions"),
        # ("job", "jobs"),
        ("procedure", "procedures"),
        ("role", "roles"),
        ("schema", "schemas"),
        ("service", "services"),
        ("secret", "secrets"),
        ("stage", "stages"),
        ("stream", "streams"),
        ("streamlit", "streamlits"),
        ("table", "tables"),
        ("task", "tasks"),
        ("user", "users"),
        ("warehouse", "warehouses"),
        ("view", "views"),
        ("image-repository", "image repositories"),
        ("git-repository", "git repositories"),
        ("notebook", "notebooks"),
    ],
)
def test_show(
    mock_connector,
    object_type,
    expected,
    mock_cursor,
    runner,
    os_agnostic_snapshot,
    mock_ctx,
):
    ctx = mock_ctx()
    mock_connector.return_value = ctx

    result = runner.invoke(["object", "list", object_type], catch_exceptions=False)
    assert result.exit_code == 0, result.output
    assert ctx.get_queries() == [f"show {expected} like '%%'"]


DESCRIBE_TEST_OBJECTS = [
    ("compute-pool", "compute_pool_example"),
    ("network-rule", "network_rule_example"),
    ("integration", "integration_example"),
    ("database", "database_example"),
    ("function", "function_example"),
    # ("job", "job_example"),
    ("procedure", "procedure_example"),
    ("role", "role_example"),
    ("schema", "schema_example"),
    ("service", "service_example"),
    ("secret", "secret_example"),
    ("stage", "stage_example"),
    ("stream", "stream_example"),
    ("streamlit", "streamlit_example"),
    ("table", "table_example"),
    ("task", "task_example"),
    ("user", "user_example"),
    ("warehouse", "warehouse_example"),
    ("view", "view_example"),
    ("git-repository", "git_repository_example"),
    ("notebook", "notebook_example"),
]


@mock.patch("snowflake.connector.connect")
@pytest.mark.parametrize(
    "object_type, input_scope, input_name",
    [
        ("schema", "database", "test_db"),
        ("table", "schema", "test_schema"),
        ("service", "compute-pool", "test_pool"),
    ],
)
def test_show_with_scope(
    mock_connector, object_type, input_scope, input_name, runner, mock_ctx
):
    ctx = mock_ctx()
    mock_connector.return_value = ctx
    result = runner.invoke(
        ["object", "list", object_type, "--in", input_scope, input_name]
    )
    obj = OBJECT_TO_NAMES[object_type]
    scope_obj = OBJECT_TO_NAMES[input_scope]
    assert result.exit_code == 0, result.output
    assert ctx.get_queries() == [
        f"show {obj.sf_plural_name} like '%%' in {scope_obj.sf_name} {input_name}"
    ]


@mock.patch("snowflake.connector.connect")
@pytest.mark.parametrize(
    "object_type, input_scope, input_name, expected",
    [
        (
            "table",
            "invalid_scope",
            "name",
            "scope must be one of the following",
        ),  # invalid scope label
        (
            "table",
            "database",
            "invalid name",
            "scope name must be a valid identifier",
        ),  # invalid scope identifier
    ],
)
def test_show_with_invalid_scope(
    mock_connector, object_type, input_scope, input_name, expected, runner
):
    result = runner.invoke(
        ["object", "list", object_type, "--in", input_scope, input_name]
    )
    assert expected in result.output


@pytest.mark.parametrize(
    "object_type, input_scope, input_name",
    [
        ("user", None, None),
        ("schema", "database", "test_db"),
        ("table", "schema", "test_schema"),
        ("service", "compute-pool", "test_pool"),
    ],
)
def test_scope_validate(object_type, input_scope, input_name):
    _scope_validate(object_type, (input_scope, input_name))


@pytest.mark.parametrize(
    "object_type, input_scope, input_name, expected_msg",
    [
        (
            "table",
            "database",
            "invalid identifier",
            "scope name must be a valid identifier",
        ),
        ("table", "invalid-scope", "identifier", "scope must be one of the following"),
        (
            "table",
            "compute-pool",
            "test_pool",
            "compute-pool scope is only supported for listing service",
        ),  # 'compute-pool' scope can only be used with 'service'
    ],
)
def test_invalid_scope_validate(object_type, input_scope, input_name, expected_msg):
    with pytest.raises(ClickException) as exc:
        _scope_validate(object_type, (input_scope, input_name))
    assert expected_msg in exc.value.message


@mock.patch("snowflake.connector")
@pytest.mark.parametrize("object_type, object_name", DESCRIBE_TEST_OBJECTS)
def test_describe(
    mock_connector, object_type, object_name, mock_cursor, runner, os_agnostic_snapshot
):
    mock_connector.connect.return_value.execute_stream.return_value = (
        None,
        mock_cursor(
            rows=[("ID", "NUMBER(38,0", "COLUMN"), ("NAME", "VARCHAR(100", "COLUMN")],
            columns=["name", "type", "kind"],
        ),
    )
    result = runner.invoke(["object", "describe", object_type, object_name])
    assert result.exit_code == 0, result.output
    assert result.output == os_agnostic_snapshot


@mock.patch("snowflake.connector")
def test_describe_fails_image_repository(mock_cursor, runner, os_agnostic_snapshot):
    result = runner.invoke(["object", "describe", "image-repository", "test_repo"])
    assert result.exit_code == 1, result.output
    assert result.output == os_agnostic_snapshot


DROP_TEST_OBJECTS = [
    *DESCRIBE_TEST_OBJECTS,
    ("image-repository", "image_repository_example"),
]


@mock.patch("snowflake.connector")
@pytest.mark.parametrize(
    "object_type, object_name",
    DROP_TEST_OBJECTS,
)
def test_drop(
    mock_connector, object_type, object_name, mock_cursor, runner, os_agnostic_snapshot
):
    mock_connector.connect.return_value.execute_stream.return_value = (
        None,
        mock_cursor(
            rows=[(f"{object_name} successfully dropped.",)],
            columns=["status"],
        ),
    )

    result = runner.invoke(["object", "drop", object_type, object_name])
    assert result.exit_code == 0, result.output
    assert result.output == os_agnostic_snapshot


@pytest.mark.parametrize("command", ["list", "drop", "describe"])
def test_that_objects_list_is_in_help(command, runner):
    result = runner.invoke(["object", command, "--help"])
    for obj in SUPPORTED_OBJECTS:
        if command == "describe" and obj == "image-repository":
            assert obj not in result.output, f"{obj} should not be in help message"
        else:
            assert obj in result.output, f"{obj} in help message"


@pytest.mark.parametrize(
    "command,expect_argument_exception",
    [
        (["object", "drop"], "OBJECT_TYPE"),
        (["object", "drop", "function"], "OBJECT_NAME"),
        (["object", "list"], "OBJECT_TYPE"),
        (["object", "describe"], "OBJECT_TYPE"),
        (["object", "describe", "function"], "OBJECT_NAME"),
    ],
)
def test_throw_exception_because_of_missing_arguments(
    runner, command, expect_argument_exception
):
    result = runner.invoke(command)
    assert result.exit_code == 2, result.output
    assert result.output.__contains__(
        f"Missing argument '{expect_argument_exception}'."
    )


def test_object_create_with_multiple_json_sources(runner):
    with NamedTemporaryFile("r") as tmp_file:
        result = runner.invoke(
            ["object", "create", "schema", "name=schema_name", "--json", "json_data"]
        )
        assert_that_result_is_usage_error(
            result,
            f"Parameters 'object_attributes' and '--json' are incompatible and cannot be used simultaneously.",
        )


def test_replace_and_not_exists_cannot_be_used_together(runner, os_agnostic_snapshot):
    result = runner.invoke(
        [
            "object",
            "create",
            "schema",
            "name=schema_name",
            "--replace",
            "--if-not-exists",
        ]
    )
    assert result.exit_code == 2, result.output
    assert result.output == os_agnostic_snapshot
