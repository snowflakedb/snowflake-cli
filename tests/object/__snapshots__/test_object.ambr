# serializer version: 1
# name: test_describe[compute-pool-compute_pool_example]
  '''
  +-----------------------------+
  | name | type        | kind   |
  |------+-------------+--------|
  | ID   | NUMBER(38,0 | COLUMN |
  | NAME | VARCHAR(100 | COLUMN |
  +-----------------------------+
  
  '''
# ---
# name: test_describe[database-database_example]
  '''
  +-----------------------------+
  | name | type        | kind   |
  |------+-------------+--------|
  | ID   | NUMBER(38,0 | COLUMN |
  | NAME | VARCHAR(100 | COLUMN |
  +-----------------------------+
  
  '''
# ---
# name: test_describe[function-function_example]
  '''
  +-----------------------------+
  | name | type        | kind   |
  |------+-------------+--------|
  | ID   | NUMBER(38,0 | COLUMN |
  | NAME | VARCHAR(100 | COLUMN |
  +-----------------------------+
  
  '''
# ---
# name: test_describe[git-repository-git_repository_example]
  '''
  +-----------------------------+
  | name | type        | kind   |
  |------+-------------+--------|
  | ID   | NUMBER(38,0 | COLUMN |
  | NAME | VARCHAR(100 | COLUMN |
  +-----------------------------+
  
  '''
# ---
# name: test_describe[integration-integration_example]
  '''
  +-----------------------------+
  | name | type        | kind   |
  |------+-------------+--------|
  | ID   | NUMBER(38,0 | COLUMN |
  | NAME | VARCHAR(100 | COLUMN |
  +-----------------------------+
  
  '''
# ---
# name: test_describe[network-rule-network_rule_example]
  '''
  +-----------------------------+
  | name | type        | kind   |
  |------+-------------+--------|
  | ID   | NUMBER(38,0 | COLUMN |
  | NAME | VARCHAR(100 | COLUMN |
  +-----------------------------+
  
  '''
# ---
# name: test_describe[notebook-notebook_example]
  '''
  +-----------------------------+
  | name | type        | kind   |
  |------+-------------+--------|
  | ID   | NUMBER(38,0 | COLUMN |
  | NAME | VARCHAR(100 | COLUMN |
  +-----------------------------+
  
  '''
# ---
# name: test_describe[procedure-procedure_example]
  '''
  +-----------------------------+
  | name | type        | kind   |
  |------+-------------+--------|
  | ID   | NUMBER(38,0 | COLUMN |
  | NAME | VARCHAR(100 | COLUMN |
  +-----------------------------+
  
  '''
# ---
# name: test_describe[role-role_example]
  '''
  +-----------------------------+
  | name | type        | kind   |
  |------+-------------+--------|
  | ID   | NUMBER(38,0 | COLUMN |
  | NAME | VARCHAR(100 | COLUMN |
  +-----------------------------+
  
  '''
# ---
# name: test_describe[schema-schema_example]
  '''
  +-----------------------------+
  | name | type        | kind   |
  |------+-------------+--------|
  | ID   | NUMBER(38,0 | COLUMN |
  | NAME | VARCHAR(100 | COLUMN |
  +-----------------------------+
  
  '''
# ---
# name: test_describe[secret-secret_example]
  '''
  +-----------------------------+
  | name | type        | kind   |
  |------+-------------+--------|
  | ID   | NUMBER(38,0 | COLUMN |
  | NAME | VARCHAR(100 | COLUMN |
  +-----------------------------+
  
  '''
# ---
# name: test_describe[service-service_example]
  '''
  +-----------------------------+
  | name | type        | kind   |
  |------+-------------+--------|
  | ID   | NUMBER(38,0 | COLUMN |
  | NAME | VARCHAR(100 | COLUMN |
  +-----------------------------+
  
  '''
# ---
# name: test_describe[stage-stage_example]
  '''
  +-----------------------------+
  | name | type        | kind   |
  |------+-------------+--------|
  | ID   | NUMBER(38,0 | COLUMN |
  | NAME | VARCHAR(100 | COLUMN |
  +-----------------------------+
  
  '''
# ---
# name: test_describe[stream-stream_example]
  '''
  +-----------------------------+
  | name | type        | kind   |
  |------+-------------+--------|
  | ID   | NUMBER(38,0 | COLUMN |
  | NAME | VARCHAR(100 | COLUMN |
  +-----------------------------+
  
  '''
# ---
# name: test_describe[streamlit-streamlit_example]
  '''
  +-----------------------------+
  | name | type        | kind   |
  |------+-------------+--------|
  | ID   | NUMBER(38,0 | COLUMN |
  | NAME | VARCHAR(100 | COLUMN |
  +-----------------------------+
  
  '''
# ---
# name: test_describe[table-table_example]
  '''
  +-----------------------------+
  | name | type        | kind   |
  |------+-------------+--------|
  | ID   | NUMBER(38,0 | COLUMN |
  | NAME | VARCHAR(100 | COLUMN |
  +-----------------------------+
  
  '''
# ---
# name: test_describe[task-task_example]
  '''
  +-----------------------------+
  | name | type        | kind   |
  |------+-------------+--------|
  | ID   | NUMBER(38,0 | COLUMN |
  | NAME | VARCHAR(100 | COLUMN |
  +-----------------------------+
  
  '''
# ---
# name: test_describe[user-user_example]
  '''
  +-----------------------------+
  | name | type        | kind   |
  |------+-------------+--------|
  | ID   | NUMBER(38,0 | COLUMN |
  | NAME | VARCHAR(100 | COLUMN |
  +-----------------------------+
  
  '''
# ---
# name: test_describe[view-view_example]
  '''
  +-----------------------------+
  | name | type        | kind   |
  |------+-------------+--------|
  | ID   | NUMBER(38,0 | COLUMN |
  | NAME | VARCHAR(100 | COLUMN |
  +-----------------------------+
  
  '''
# ---
# name: test_describe[warehouse-warehouse_example]
  '''
  +-----------------------------+
  | name | type        | kind   |
  |------+-------------+--------|
  | ID   | NUMBER(38,0 | COLUMN |
  | NAME | VARCHAR(100 | COLUMN |
  +-----------------------------+
  
  '''
# ---
# name: test_describe_fails_image_repository
  '''
  +- Error ----------------------------------------------------------------------+
  | Describe is currently not supported for object of type image-repository      |
  +------------------------------------------------------------------------------+
  
  '''
# ---
# name: test_drop[compute-pool-compute_pool_example]
  '''
  +--------------------------------------------+
  | status                                     |
  |--------------------------------------------|
  | compute_pool_example successfully dropped. |
  +--------------------------------------------+
  
  '''
# ---
# name: test_drop[database-database_example]
  '''
  +----------------------------------------+
  | status                                 |
  |----------------------------------------|
  | database_example successfully dropped. |
  +----------------------------------------+
  
  '''
# ---
# name: test_drop[function-function_example]
  '''
  +----------------------------------------+
  | status                                 |
  |----------------------------------------|
  | function_example successfully dropped. |
  +----------------------------------------+
  
  '''
# ---
# name: test_drop[git-repository-git_repository_example]
  '''
  +----------------------------------------------+
  | status                                       |
  |----------------------------------------------|
  | git_repository_example successfully dropped. |
  +----------------------------------------------+
  
  '''
# ---
# name: test_drop[image-repository-image_repository_example]
  '''
  +------------------------------------------------+
  | status                                         |
  |------------------------------------------------|
  | image_repository_example successfully dropped. |
  +------------------------------------------------+
  
  '''
# ---
# name: test_drop[integration-integration_example]
  '''
  +-------------------------------------------+
  | status                                    |
  |-------------------------------------------|
  | integration_example successfully dropped. |
  +-------------------------------------------+
  
  '''
# ---
# name: test_drop[network-rule-network_rule_example]
  '''
  +--------------------------------------------+
  | status                                     |
  |--------------------------------------------|
  | network_rule_example successfully dropped. |
  +--------------------------------------------+
  
  '''
# ---
# name: test_drop[notebook-notebook_example]
  '''
  +----------------------------------------+
  | status                                 |
  |----------------------------------------|
  | notebook_example successfully dropped. |
  +----------------------------------------+
  
  '''
# ---
# name: test_drop[procedure-procedure_example]
  '''
  +-----------------------------------------+
  | status                                  |
  |-----------------------------------------|
  | procedure_example successfully dropped. |
  +-----------------------------------------+
  
  '''
# ---
# name: test_drop[role-role_example]
  '''
  +------------------------------------+
  | status                             |
  |------------------------------------|
  | role_example successfully dropped. |
  +------------------------------------+
  
  '''
# ---
# name: test_drop[schema-schema_example]
  '''
  +--------------------------------------+
  | status                               |
  |--------------------------------------|
  | schema_example successfully dropped. |
  +--------------------------------------+
  
  '''
# ---
# name: test_drop[secret-secret_example]
  '''
  +--------------------------------------+
  | status                               |
  |--------------------------------------|
  | secret_example successfully dropped. |
  +--------------------------------------+
  
  '''
# ---
# name: test_drop[service-service_example]
  '''
  +---------------------------------------+
  | status                                |
  |---------------------------------------|
  | service_example successfully dropped. |
  +---------------------------------------+
  
  '''
# ---
# name: test_drop[stage-stage_example]
  '''
  +-------------------------------------+
  | status                              |
  |-------------------------------------|
  | stage_example successfully dropped. |
  +-------------------------------------+
  
  '''
# ---
# name: test_drop[stream-stream_example]
  '''
  +--------------------------------------+
  | status                               |
  |--------------------------------------|
  | stream_example successfully dropped. |
  +--------------------------------------+
  
  '''
# ---
# name: test_drop[streamlit-streamlit_example]
  '''
  +-----------------------------------------+
  | status                                  |
  |-----------------------------------------|
  | streamlit_example successfully dropped. |
  +-----------------------------------------+
  
  '''
# ---
# name: test_drop[table-table_example]
  '''
  +-------------------------------------+
  | status                              |
  |-------------------------------------|
  | table_example successfully dropped. |
  +-------------------------------------+
  
  '''
# ---
# name: test_drop[task-task_example]
  '''
  +------------------------------------+
  | status                             |
  |------------------------------------|
  | task_example successfully dropped. |
  +------------------------------------+
  
  '''
# ---
# name: test_drop[user-user_example]
  '''
  +------------------------------------+
  | status                             |
  |------------------------------------|
  | user_example successfully dropped. |
  +------------------------------------+
  
  '''
# ---
# name: test_drop[view-view_example]
  '''
  +------------------------------------+
  | status                             |
  |------------------------------------|
  | view_example successfully dropped. |
  +------------------------------------+
  
  '''
# ---
# name: test_drop[warehouse-warehouse_example]
  '''
  +-----------------------------------------+
  | status                                  |
  |-----------------------------------------|
  | warehouse_example successfully dropped. |
  +-----------------------------------------+
  
  '''
# ---
# name: test_replace_and_not_exists_cannot_be_used_together
  '''
  Usage: root object create [OPTIONS] OBJECT_TYPE [OBJECT_ATTRIBUTES]...
  Try 'root object create --help' for help.
  +- Error ----------------------------------------------------------------------+
  | Parameters '--if-not-exists' and '--replace' are incompatible and cannot be  |
  | used simultaneously.                                                         |
  +------------------------------------------------------------------------------+
  
  '''
# ---
